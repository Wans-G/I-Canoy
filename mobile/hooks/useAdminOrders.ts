// src/hooks/useAdminOrders.ts
import { useState, useCallback } from 'react'
import AsyncStorage from '@react-native-async-storage/async-storage'

export type OrderStatus = 'Out for Delivery' | 'Processing' | 'Canceled'

export interface OrderDetail {
  id: number
  stripe_payment_method_id: string | null
  shipping_address: string
  city: string
  state: string
  zip_code: string
  total_price: string
  status: string
  tracking_number: string | null
  created_at: string
  user: { first_name: string; last_name: string; email: string }
  items: { product_name: string; product_price: string; quantity: number }[]
}

type ApiResponse = {
  results: Omit<OrderDetail, 'status'> & { status: string }[]
  next: any
}

export function useAdminOrders(apiBase: string) {
  const [orders, setOrders] = useState<Record<OrderStatus, OrderDetail[]>>({
    'Out for Delivery': [],
    Processing: [],
    Canceled: [],
  })
  const [loading, setLoading] = useState(false)
  const [hasNext, setHasNext] = useState(true)

  const fetchOrders = useCallback(
    async (pageNum = 1) => {
      const token = await AsyncStorage.getItem('authToken')
      if (!token) return null

      setLoading(true)
      const res = await fetch(
        `${apiBase}/api/store/orders/admin/?page=${pageNum}`,
        { headers: { Authorization: `Token ${token}` } }
      )
      const data: ApiResponse = await res.json()
      setLoading(false)
      setHasNext(Boolean(data.next))

      // group by our friendly labels
      const grouped: Record<OrderStatus, OrderDetail[]> = {
        'Out for Delivery': [],
        Processing: [],
        Canceled: [],
      }

      for (const o of data.results) {
        const key: OrderStatus =
          o.status === 'out_for_delivery'
            ? 'Out for Delivery'
            : o.status === 'cancelled'
            ? 'Canceled'
            : 'Processing'
        grouped[key].push({ ...o, status: o.status })
      }

      setOrders(prev =>
        pageNum === 1
          ? grouped
          : {
              'Out for Delivery': [...prev['Out for Delivery'], ...grouped['Out for Delivery']],
              Processing: [...prev.Processing, ...grouped.Processing],
              Canceled: [...prev.Canceled, ...grouped.Canceled],
            }
      )

      return data
    },
    [apiBase]
  )

  const completeOrder = useCallback(
    async (orderId: number, tracking: string) => {
      const token = await AsyncStorage.getItem('authToken')
      if (!token) throw new Error('no token')

      const res = await fetch(
        `${apiBase}/api/store/orders/update/${orderId}/`,
        {
          method: 'POST',
          headers: {
            Authorization: `Token ${token}`,
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            status: 'out_for_delivery',
            tracking_number: tracking,
          }),
        }
      )
      if (!res.ok) throw new Error('update failed')
      const updated: OrderDetail = await res.json()

      // move into Out for Delivery
      setOrders(prev => {
        const copy = { ...prev }
        // remove from Processing
        copy.Processing = copy.Processing.filter(o => o.id !== updated.id)
        // add to Out for Delivery
        copy['Out for Delivery'] = [
          ...copy['Out for Delivery'],
          { ...updated, status: 'out_for_delivery' },
        ]
        return copy
      })
      return updated
    },
    [apiBase]
  )

  return { orders, loading, hasNext, fetchOrders, completeOrder }
}
