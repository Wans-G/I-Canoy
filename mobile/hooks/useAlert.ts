import { Alert, Platform } from "react-native";
import Toast from "react-native-toast-message";

const showMessage = () => {
  const useAlert = (title: string, message?: string) => {
        Alert.alert(title, message);   
    }
  const useToast = (title: string, message?: string) => {
      Toast.show({
        type: "success",
        text1: title,
        text2: message || "",
        position: "top",
        topOffset: Platform.OS === "web" ? 20 : 70,
        visibilityTime: 4000,
        autoHide: true,
      });
    }
  return { useAlert, useToast };

};

export default showMessage;
