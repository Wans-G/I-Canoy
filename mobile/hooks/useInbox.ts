import { collection, query, where, onSnapshot, doc, getDoc } from "firebase/firestore";
import { firestore } from "../_utlis/firebaseConfig";
import { useEffect, useState } from "react";

type InboxItem = {
  id: string;
  type: "dm" | "group";
  name: string;
  profilePicture: string;
  lastMessage: string;
  lastUpdated: string;
};

export const useInbox = (currentUserId: number | null) => {
  const [inbox, setInbox] = useState<InboxItem[]>([]);

  useEffect(() => {
    if (!currentUserId) return;
    const numericId = Number(currentUserId);

    const dmQuery = query(
      collection(firestore, "conversations"),
      where("members", "array-contains", numericId)
    );

    const groupQuery = query(
      collection(firestore, "groupChats"),
      where("members", "array-contains", numericId)
    );

    const unsubscribeDMs = onSnapshot(dmQuery, async (snapshot) => {
      const enrichedDMs = await Promise.all(snapshot.docs.map(async (docSnap) => {
        const data = docSnap.data();
        const otherId = data.members.find((id: number) => id !== numericId);

        const userDoc = await getDoc(doc(firestore, "users", otherId.toString()));
        console.log("ðŸ“¦ Fetching user doc from Firestore for ID:", otherId.toString());
        const userData = userDoc.exists() ? userDoc.data() : { username: "Unknown", profilePicture: "" };


            return {
          id: docSnap.id,
          type: "dm" as const,
          name: userData.username,
          profilePicture: userData.profilePicture,
          lastMessage: data.lastMessage,
          lastUpdated: data.lastUpdated?.toDate().toISOString() || new Date().toISOString(),
        };
      }));

      setInbox((prev) =>
        mergeAndSort([...prev.filter(i => i.type !== "dm"), ...enrichedDMs])
      );
    });

    const unsubscribeGroups = onSnapshot(groupQuery, (snapshot) => {
      const groups = snapshot.docs.map((doc) => ({
        id: doc.id,
        type: "group" as const,
        name: doc.data().name,
        profilePicture: "", // Add group profile picture support if needed
        lastMessage: doc.data().lastMessage,
        lastUpdated: doc.data().lastUpdated?.toDate().toISOString() || new Date().toISOString(),
      }));
      setInbox((prev) =>
        mergeAndSort([...prev.filter(i => i.type !== "group"), ...groups])
      );
    });

    return () => {
      unsubscribeDMs();
      unsubscribeGroups();
    };
  }, [currentUserId]);

  return inbox;
};

function mergeAndSort(items: InboxItem[]): InboxItem[] {
  return items.sort((a, b) => new Date(b.lastUpdated).getTime() - new Date(a.lastUpdated).getTime());
}
