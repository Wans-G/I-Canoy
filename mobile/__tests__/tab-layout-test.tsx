/**
 * __tests__/TabLayout.test.tsx
 */
import React from 'react'
import { render } from '@testing-library/react-native'

// 1) Mock useColorScheme hook
jest.mock('@/hooks/useColorScheme', () => ({
  useColorScheme: jest.fn(),
}))
// 2) Mock Colors
jest.mock('@/constants/Colors', () => ({
  Colors: {
    light: { tint: 'lightTint' },
    dark: { tint: 'darkTint' },
  },
}))
// 3) Mock HapticTab
jest.mock('@/components/HapticTab', () => ({
  HapticTab: () => null,
}))
// 4) Mock TabBarBackground
jest.mock('@/components/ui/TabBarBackground', () => () => null)
// 5) Mock IconSymbol
jest.mock('@/components/ui/IconSymbol', () => {
  const React = require('react')
  return {
    IconSymbol: (props: any) => React.createElement('IconSymbol', props),
  }
})
// 6) Mock expo-routerâ€™s Tabs
jest.mock('expo-router', () => {
  const React = require('react')
  const { View } = require('react-native')

  function Tabs(props: any) {
    return React.createElement(React.Fragment, null, props.children)
  }
  Tabs.Screen = ({ name }: { name: string }) =>
    React.createElement(View, { testID: `screen-${name}` })

  return { Tabs }
})

// 7) Import under test
import TabLayout from '../app/(tabs)/_layout'
import { useColorScheme } from '@/hooks/useColorScheme'

describe('TabLayout', () => {
  beforeEach(() => {
    ;(useColorScheme as jest.Mock).mockReturnValue('light')
  })

  it('renders all four tab screens', () => {
    const { getByTestId } = render(<TabLayout />)

    expect(getByTestId('screen-home')).toBeTruthy()
    expect(getByTestId('screen-store')).toBeTruthy()
    expect(getByTestId('screen-social')).toBeTruthy()
    expect(getByTestId('screen-profile')).toBeTruthy()
  })
})
