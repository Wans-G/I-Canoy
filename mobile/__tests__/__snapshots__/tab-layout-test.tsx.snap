// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabLayout matches snapshot with dark tint 1`] = `
[
  <View
    accessibilityLabel="Home"
    testID="screen-home"
  />,
  <View
    accessibilityLabel="Store"
    testID="screen-store"
  />,
  <View
    accessibilityLabel="Social"
    testID="screen-social"
  />,
  <View
    accessibilityLabel="Profile"
    testID="screen-profile"
  />,
]
`;

exports[`TabLayout matches snapshot with light tint 1`] = `
[
  <View
    accessibilityLabel="Home"
    testID="screen-home"
  />,
  <View
    accessibilityLabel="Store"
    testID="screen-store"
  />,
  <View
    accessibilityLabel="Social"
    testID="screen-social"
  />,
  <View
    accessibilityLabel="Profile"
    testID="screen-profile"
  />,
]
`;
